---
title: "First comparison"
author: "gg"
date: '`r format(Sys.time(), "%d %B, %Y")`'
bibliography: /Users/ggloor/Library/texmf/bibtex/bib/bibdesk_refs.bib
fig_caption: true
output:
    pdf_document:
        fig_caption: yes
---

To run this file:
Rscript -e "rmarkdown::render('ALDEx_comparison.Rmd')"
## Types of data

Comparison of 'differential abundance' is problematic for compositional data [@fernandes:2013;@fernandes:2014]. Since the apparent abundance of every value depends on the apparent abundance of every other value, we can get into real difficulties if we are not careful. Take the simple example where we have two samples. The samples contain the following counts for five features:

\begin{math}
A = [1000, 100, 50, 250]
\end{math} and
\begin{math}
B=[10, 500, 250, 1250]
\end{math}.

We want to answer the question: Have the abundances of the features changed?

We sequence, and have a total count of about 100 (it is a first generation machine!)

So we get:
\begin{math}A_s = [71,7,4,18], B_s = [1,25,12,62]
\end{math}

Note that these values appear to be very different between the groups. However, if we take one feature as a reference, say feature 4, and determine a ratio, i.e.:

\begin{math}
A_r = [ 74/18, 7/18, 4/18 ] = [ 4.1, 0.39, 0.22 ]
\end{math}

\begin{math}
B_r = [ 1/62, 25/62, 12/62 ] = [ 0.02 , 0.40, 0.20 ]
\end{math}

Here we can see that if we assume one feature is constant (feature 4), then the last two are seen to be very similar in abundance. Now we can infer that the majority of change is in the first feature. We cannot compare the last feature because it is assumed to be constant, that is, the assumed change in the last feature is 0. This approach is the one used by ANCOM, a recently developed tool to assess change in microbiome datasets [@ancom:2015].

Since we cannot know which feature, if any, is constant, we can assume that a large number of the features exhibit only random change.  Then rather than using one feature as a reference we can use the geometric mean abundance of all features. Note: this approach works poorly if there are only a small number of features (less than about 50) or if the features are very asymmetrically distributed between groups. This approach is the one used by ALDEx2 [@fernandes:2013;@fernandes:2014], and is the method that we will use.

One complication is that a genometric mean cannot be determined if any of the values have a count of 0. For pairwise comparisons


```{r, echo=TRUE,eval=TRUE}
library(ALDEx2)

# read the dataset
d <- read.table("data/barton_agg.tsv", row.names=1, header=T)
# make a vector containing the two names of the conditions
# in the same order as in the column names

d.conds <- c(rep("SNF", length(grep("SNF", rownames(d))) ),
    rep("WT", length(grep("WT", rownames(d)))) )


# generate Monte-Carlo instances of the probability of observing each count
# given the actual read count and the observed read count.
# use a prior of 0.5, corresponding to maximal uncertainty about the read count
# this returns a set of clr values, one for each mc instance
# this workflow can take several minutes

# note that the latest version of ALDEx2 requires conditions explicitly
d.x <- aldex.clr(t(d), conds=d.conds, mc.samples=128)

# calculate effect sizes for each mc instance, report the expected value
d.eff <- aldex.effect(d.x, d.conds, include.sample.summary=TRUE)

# perform parametric or non-parametric tests for difference
# report the expected value of the raw and BH-corrected P value
d.tt <- aldex.ttest(d.x, d.conds)

# concatenate everything into one file
x.all <- data.frame(d.eff,d.tt)
```

We will display the results using a number of different plots to show how each plot gives a different way of exploring the data. The mainstay that we advocate is the effect plot [@Gloor:2015], that plots the constituents of normalized change, or effect size.

```{r, echo=FALSE,eval=TRUE, results='as.is', fig.width=7, fig.height=7, error=FALSE, message=FALSE, warning=FALSE, fig.cap="Plotted here are features with no difference between groups (grey), a statistically difference between groups (red), and with an effect larger than 2 (blue circles). These are plotted using different plots (described clockwise from top left). The effect plot [@Gloor:2015] illustrates the difference between groups vs. the dispersion (variance) within groups. If the effect is greater than one (outside the grey lines), then, on average the features are obviosly separable by eye when plotted; roughly, they would be seen to have a greater difference between groups than the pooled standard deviation. Effect is a more robust measure of difference than are P values, since the latter depend on sample size; large sample sizes will always give low P values [@Halsey:2015aa]. We can see this here where the large sample size means that even highly variable OTUs are significantly different. The Bland-Altman plot [@altman:1983] compares difference and abundance, and is often seen in RNA-Seq data. The Volcano plot [@Cui:2003aa] shows the association between difference and P value, and the final plot shows the association between effect and P value."}

# x.all <- read.table("data/barton_agg.tsv", header=T, row.names=1)

# get 'significant' set
sig <- x.all$wi.eBH < 0.05
eff <- abs(x.all$effect) > 2

# plot all in transparent grey
# low BH-corrected p values as red
# effect sizes > 2 as blue+red
par(mfrow=c(2,2))

plot(x.all$diff.win, x.all$diff.btw, col=rgb(0,0,0,0.3), pch=19,
    cex=0.5, ylim=c(-6,6), xlim=c(0,6), xlab="dispersion", ylab="difference",
    main="Effect plot")
points(x.all$diff.win[sig], x.all$diff.btw[sig], col=rgb(1,0,0,0.3), pch=19, cex=0.5 )
points(x.all$diff.win[eff], x.all$diff.btw[eff], col=rgb(0,0,1,0.6), pch=21, cex=0.7 )
abline(0,1, lty=2, lwd=2, col=rgb(0,0,0,0.4))
abline(0,-1, lty=2, lwd=2, col=rgb(0,0,0,0.4))

plot(x.all$rab.all, x.all$diff.btw, col=rgb(0,0,0,0.3), pch=19,
    cex=0.5, ylim=c(-6,6), xlab="clr abundance", ylab="difference",
    main="Bland-Altman plot")
points(x.all$rab.all[sig], x.all$diff.btw[sig], col=rgb(1,0,0,0.3), pch=19, cex=0.5 )
points(x.all$rab.all[eff], x.all$diff.btw[eff], col=rgb(0,0,1,0.6), pch=21, cex=0.7 )

plot(x.all$diff.btw, x.all$wi.ep, col=rgb(0,0,0,0.3), pch=19,
    cex=0.5, xlab="difference", ylab="log p value",
    main="Difference vs. p plot", log="y")
points(x.all$diff.btw[sig], x.all$wi.ep[sig], col=rgb(1,0,0,0.3), pch=19, cex=0.5 )
points(x.all$diff.btw[eff], x.all$wi.ep[eff], col=rgb(0,0,1,0.6), pch=21, cex=0.7 )

plot(x.all$effect, x.all$wi.ep, col=rgb(0,0,0,0.3), pch=19,
    cex=0.5,  xlab="effect", ylab="log p value",
    main="Effect vs. p plot", log="y")
points(x.all$effect[sig], x.all$wi.ep[sig], col=rgb(1,0,0,0.3), pch=19, cex=0.5 )
points(x.all$effect[eff], x.all$wi.ep[eff], col=rgb(0,0,1,0.6), pch=21, cex=0.7 )

```

\newpage

The effect sizes can be understood as a measure of separability between groups for each feature. Plotted in the figure are features with different effect sizes and the corresponding adjusted p value is included. In RNA-seq data, we have found that an effect size cutoff between 1 and 2 [@macklaim:2013] gives reliable results in meta-transcriptome analyses.

```{r, echo=FALSE,eval=TRUE, results='as.is', fig.width=7, fig.height=7, error=FALSE, message=FALSE, warning=FALSE, fig.cap="Histograms showing the separation between groups when choosing features withdiffering effect sizes. Features  with the largest effect are the most reliably different between groups, and should be chosen over those that are most significantly different whenever possible. Note that an effect size of 0.5 is more than sufficient to give a significant difference but the separation between groups is marginal."}

par(mfrow=c(2,2))

hist(as.numeric(x.all["PWR1",4:51]), breaks=11, xlim=c(-8,-2), col=rgb(1,0,0,0.3),
    xlab="rAB PWR1", main="Effect: 0.5 wi.eBH: 4e-4", ylim=c(0,12) )
hist(as.numeric(x.all["PWR1",52:99]), add=T,  breaks=20, col=rgb(0,0,1,0.3))

hist(as.numeric(x.all["NTS2.1",4:51]), breaks=20, col=rgb(1,0,0,0.3),
    xlab="rAB NTS2.1", main="Effect: 1.1 wi.eBH: 2e-11", ylim=c(0,15) )
hist(as.numeric(x.all["NTS2.1",52:99]), add=T,  breaks=11, col=rgb(0,0,1,0.3))

hist(as.numeric(x.all["YPR145W",4:51]), xlim=c(2,6), breaks=11, col=rgb(1,0,0,0.3),
    xlab="rAB YPR145W", main="Effect: 2 wi.eBH: 6.5e-26", ylim=c(0,12) )
hist(as.numeric(x.all["YPR145W",52:99]), add=T,  breaks=20, col=rgb(0,0,1,0.3))

hist(as.numeric(x.all["YOR290C",4:51]), xlim=c(-11,5), breaks=20, col=rgb(1,0,0,0.3),
    xlab="rAB YOR290C", main="Effect: 5.6 wi.eBH: 1e-26", ylim=c(0,40) )
hist(as.numeric(x.all["YOR290C",52:99]), add=T,  breaks=3, col=rgb(0,0,1,0.3))
```


\newpage
#References


